akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -778,18 +778,25 @@ private[akka] class ShardRegion(
   }
 
   def receiveCommand(cmd: ShardRegionCommand): Unit = cmd match {
+
+    // this is periodic tick to run some house-keeping
     case Retry =>
-      sendGracefulShutdownToCoordinator()
 
-      if (shardBuffers.nonEmpty)
-        retryCount += 1
-      if (coordinator.isEmpty)
-        register()
-      else {
-        requestShardBufferHomes()
-      }
+      // retryCount is used to avoid flooding the logs
+      // it's used inside register() and requestShardBufferHomes() whenever shardBuffers.nonEmpty
+      // therefore we update it if needed on each Retry msg
+      if (shardBuffers.nonEmpty) retryCount += 1
+
+      // we depend on the coordinator each time, if empty we need to register
+      // otherwise we can try to deliver some buffered messages
+      if (coordinator.isEmpty) register()
+      else tryRequestShardBufferHomes()
 
-      tryCompleteGracefulShutdown()
+      // eventually, also re-trigger a graceful shutdown if one is in progress
+      // Note: we do try to deliver buffered messages even in the middle of a graceful shutdown
+      // every message that we manage to deliver is a win
+      sendGracefulShutdownToCoordinatorIfInProgress()
+      tryCompleteGracefulShutdownIfInProgress()
```

Comment:
I added some comments to clarify why we are doing it. 

Note that I moved the `sendGracefulShutdownToCoordinatorIfInProgress` to underneath. It doesn't matter much when we do it, AFAIC.

Let me know if I'm missing something here.

Reply:
good adjustments